stdlib/src/collections/string.mojo

```diff
@@ -1613,47 +1622,50 @@ struct String(
         .
         """
 
-        fn num_bytes(b: UInt8) -> Int:
-            var flipped = ~b
-            return int(count_leading_zeros(flipped) + (flipped >> 7))
-
         var output = List[String]()
         var str_byte_len = self.byte_length() - 1
         var lhs = 0
         var rhs = 0
         var items = 0
+        var ptr = self.unsafe_ptr()
+        alias SelfSlice = StringSlice[__lifetime_of(self)]
+
+        @always_inline("nodebug")
+        fn _build_slice(s_ptr: UnsafePointer[UInt8], length: Int) -> SelfSlice:
```

Comment:
**Question** Have you considered having this private function work on a `Span` instead of continuing with ptr + length?  I haven't checked to see if `StringSlice` has a constructor from a `Span`, but if not, we could pull the ptr + length out of the span when needed.  It helps tighten up the types until we absolutely need to work with ptr + length bare.  WDYT?

Reply:
yes it has a constructor, will do. It looks like it will simplify things :D

Reply:
Only problem that I now realize is that using span for indexing will hurt performance because of index normalization. I'll revert to using pointers for indexing since this function is pretty performance sensitive. There is also the fact that doing `String(StringSlice(Span(ptr, len)))` seems a bit redundant when there is `String(StringSlice(ptr, len))`.